core/clingutils/src/TClingUtils.cxx

```diff
@@ -1473,6 +1474,7 @@ static void CreateNameTypeMap(const clang::CXXRecordDecl &cl, ROOT::MembersTypeM
    std::string typenameStr;
 
    const clang::ASTContext& astContext =  cl.getASTContext();
+   auto CTSD = llvm::dyn_cast<clang::ClassTemplateSpecializationDecl>(&cl);
```

Comment:
Looks like this example breaks some assumptions in this PR.

```cpp
#include <vector>
#include <string>
#include <stdint.h>

template <class T>
class MEtoEDM {
public:
  typedef std::vector<uint32_t> TagList;

  struct MEtoEDMObject {
    std::string name;
    TagList tags;
    T object;
  };
};
```

If you can reproduce that crash with a test using the example that'd be a good step forward.

```
#0  0x00007fffe88f652f in raise () from /lib64/libc.so.6
#1  0x00007fffe88c9e65 in abort () from /lib64/libc.so.6
#2  0x00007fffe88c9d39 in __assert_fail_base.cold.0 () from /lib64/libc.so.6
#3  0x00007fffe88eee86 in __assert_fail () from /lib64/libc.so.6
#4  0x00007fffeaef3d10 in cling::utils::CreateNestedNameSpecifierForScopeOf (Ctx=..., decl=0x3d176f8, FullyQualified=true, Spec=0x0)
    at /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/el8_amd64_gcc12/lcg/root/6.33.01-48fd78c793bc73bd3f87499c5ca7bf90/root-6.33.01/interpreter/cling/lib/Utils/AST.cpp:1561
#5  0x00007fffeaef3f03 in cling::utils::CreateNestedNameSpecifierForScopeOf (Ctx=..., TypePtr=0x3d17a30, FullyQualified=true, 
    Spec=0x0)
    at /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/el8_amd64_gcc12/lcg/root/6.33.01-48fd78c793bc73bd3f87499c5ca7bf90/root-6.33.01/interpreter/cling/lib/Utils/AST.cpp:1621
#6  0x00007fffeaef4573 in cling::utils::TypeName::GetFullyQualifiedType (QT=..., Ctx=..., Spec=0x0)
    at /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/el8_amd64_gcc12/lcg/root/6.33.01-48fd78c793bc73bd3f87499c5ca7bf90/root-6.33.01/interpreter/cling/lib/Utils/AST.cpp:1760
#7  0x00007fffeaef478e in cling::utils::TypeName::GetFullyQualifiedName[abi:cxx11](clang::QualType, clang::ASTContext const&, clang::ClassTemplateSpecializationDecl const*) (QT=..., Ctx=..., Spec=0x0)
    at /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/el8_amd64_gcc12/lcg/root/6.33.01-48fd78c793bc73bd3f87499c5ca7bf90/root-6.33.01/interpreter/cling/lib/Utils/AST.cpp:1805
#8  0x00007fffea825603 in ROOT::TMetaUtils::GetFullyQualifiedTypeName (typenamestr="", qtype=..., astContext=..., Spec=0x0)
    at /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/el8_amd64_gcc12/lcg/root/6.33.01-48fd78c793bc73bd3f87499c5ca7bf90/root-6.33.01/core/clingutils/src/TClingUtils.cxx:3497
#9  0x00007fffea81c1ef in CreateNameTypeMap (cl=..., nameType=std::map with 1 element = {...})
    at /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/el8_amd64_gcc12/lcg/root/6.33.01-48fd78c793bc73bd3f87499c5ca7bf90/root-6.33.01/core/clingutils/src/TClingUtils.cxx:1502
#10 0x00007fffea81d7a2 in ROOT::TMetaUtils::WriteClassInit (finalString=..., cl=..., decl=0x3d5f6c8, interp=..., normCtxt=..., 
    ctorTypes=std::__cxx11::list = {...}, needCollectionProxy=@0x7ffffffef3d7: false)
    at /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/el8_amd64_gcc12/lcg/root/6.33.01-48fd78c793bc73bd3f87499c5ca7
```

That is, we introduce a `ClassTemplateSpecializationDecl` which could be nullptr and yet we will still have a class template. In this line here we get:

```
(gdb) p cl.getDescribedTemplate()
$15 = (clang::TemplateDecl *) 0x0
(gdb) p cl.getDescribedClassTemplate()
$16 = (clang::ClassTemplateDecl *) 0x0
(gdb) p cl.getTemplateSpecializationKind () 
$17 = clang::TSK_ImplicitInstantiation
```

`p cl.getInstantiatedFromMemberClass ()` will get us the outer class will still not be the class template specialization decl.

I think we are trying to get the template pattern of the outer class asking its nested class and that's probably not a good idea. I am not entirely sure what this pull request intends to do, looks very hacky anyways, you can probably get away by asking if the class is nested.

I think this PR is quite messy and will require a lot of work to make it consistent and mergeable.